File: app/schemas/evaluation.py
from pydantic import BaseModel, ConfigDict
from datetime import date
from app.models.user import UserRole
from app.models.evaluation import EvaluationItem
from typing import List, Optional

# EvaluationWeight Schemas
class EvaluationWeightBase(BaseModel):
    role: UserRole
    item: EvaluationItem
    weight: float

class EvaluationWeightCreate(EvaluationWeightBase):
    pass

class EvaluationWeightUpdate(EvaluationWeightBase):
    pass

class EvaluationWeightInDB(EvaluationWeightBase):
    id: int

    model_config = ConfigDict(from_attributes=True)

class EvaluationWeight(EvaluationWeightInDB):
    pass

# PeerEvaluation Schemas
class PeerEvaluationBase(BaseModel):
    project_id: int
    evaluatee_id: int
    score: int
    comment: Optional[str] = None

class PeerEvaluationCreate(BaseModel):
    evaluations: List[PeerEvaluationBase]

class PeerEvaluationInDBBase(PeerEvaluationBase):
    id: int
    evaluator_id: int
    evaluation_period: str

    model_config = ConfigDict(from_attributes=True)

class PeerEvaluation(PeerEvaluationInDBBase):
    pass

# PmEvaluation Schemas
class PmEvaluationBase(BaseModel):
    project_id: int
    evaluatee_id: int
    score: int
    comment: Optional[str] = None

class PmEvaluationCreate(BaseModel):
    evaluations: List[PmEvaluationBase]

class PmEvaluationInDBBase(PmEvaluationBase):
    id: int
    evaluator_id: int
    evaluation_period: str

    model_config = ConfigDict(from_attributes=True)

class PmEvaluation(PmEvaluationInDBBase):
    pass


# Schema for Admin to create a single PM evaluation for a PM
class PmSelfEvaluationCreate(PmEvaluationBase):
    pass


# QualitativeEvaluation Schemas
class QualitativeEvaluationBase(BaseModel):
    evaluatee_id: int
    score: int
    feedback: Optional[str] = None


class QualitativeEvaluationCreate(BaseModel):
    evaluations: List[QualitativeEvaluationBase]


class QualitativeEvaluationInDBBase(QualitativeEvaluationBase):
    id: int
    evaluator_id: int
    evaluation_period: str

    model_config = ConfigDict(from_attributes=True)


class QualitativeEvaluation(QualitativeEvaluationInDBBase):
    pass


# FinalEvaluation Schemas
class FinalEvaluationBase(BaseModel):
    evaluatee_id: int
    evaluation_period: str
    peer_score: float | None = None
    pm_score: float | None = None
    qualitative_score: float | None = None
    final_score: float
    grade: Optional[str] = None


class FinalEvaluationCreate(FinalEvaluationBase):
    pass


class FinalEvaluationUpdate(FinalEvaluationBase):
    pass


class FinalEvaluationInDB(FinalEvaluationBase):
    id: int

    model_config = ConfigDict(from_attributes=True)


class FinalEvaluation(FinalEvaluationInDB):
    pass

# Request body for calculate_final_evaluations endpoint
class FinalEvaluationCalculateRequest(BaseModel):
    user_ids: Optional[List[int]] = None

# EvaluationPeriod Schemas
class EvaluationPeriodBase(BaseModel):
    name: str
    start_date: date
    end_date: date

class EvaluationPeriodCreate(EvaluationPeriodBase):
    pass

class EvaluationPeriodUpdate(EvaluationPeriodBase):
    pass

class EvaluationPeriodInDB(EvaluationPeriodBase):
    id: int

    model_config = ConfigDict(from_attributes=True)

class EvaluationPeriod(EvaluationPeriodInDB):
    is_active: Optional[bool] = None


# DepartmentGradeRatio Schemas
class DepartmentGradeRatioBase(BaseModel):
    department_grade: str
    s_ratio: float
    a_ratio: float

class GradeAdjustment(BaseModel):
    user_id: int
    grade: str

class GradeAdjustmentRequest(BaseModel):
    evaluation_period: str
    adjustments: List[GradeAdjustment]

class DepartmentGradeRatioCreate(DepartmentGradeRatioBase):
    pass

class DepartmentGradeRatioUpdate(DepartmentGradeRatioBase):
    pass

class DepartmentGradeRatioInDB(DepartmentGradeRatioBase):
    id: int

    model_config = ConfigDict(from_attributes=True)


class DepartmentGradeRatio(DepartmentGradeRatioInDB):
    pass


# Schemas for viewing evaluation results
class PmScoreResult(BaseModel):
    project_name: str
    pm_name: str
    score: int


class MyEvaluationResult(BaseModel):
    evaluation_period: str
    grade: Optional[str] = None
    pm_scores: List[PmScoreResult]


class ManagerEvaluationView(BaseModel):
    final_evaluation: FinalEvaluation
    peer_feedback: List[str]

    model_config = ConfigDict(from_attributes=True)


# Schemas for Evaluation UX Improvement
class MyEvaluationTask(BaseModel):
    project_id: int
    project_name: str
    user_role_in_project: str # "PM" or "MEMBER"


class PeerEvaluationTarget(BaseModel):
    evaluatee_id: int
    evaluatee_name: str
    score: Optional[int] = None
    comment: Optional[str] = None


class PeerEvaluationDetail(BaseModel):
    project_id: int
    project_name: str
    status: str # "NOT_STARTED", "IN_PROGRESS", "COMPLETED"
    peers_to_evaluate: List[PeerEvaluationTarget]


class PmEvaluationTarget(BaseModel):
    evaluatee_id: int
    evaluatee_name: str
    score: Optional[int] = None
    comment: Optional[str] = None


class PmEvaluationDetail(BaseModel):
    project_id: int
    project_name: str
    status: str # "NOT_STARTED", "IN_PROGRESS", "COMPLETED"
    members_to_evaluate: List[PmEvaluationTarget]

File: frontend/src/schemas/evaluation.ts
export interface EvaluationPeriod {
    id: number;
    name: string;
    start_date: string;
    end_date: string;
    is_active: boolean;
  }
  
  export interface EvaluationPeriodCreate {
      name: string;
      start_date: string;
      end_date: string;
  }
  
  export type EvaluationPeriodUpdate = Partial<EvaluationPeriodCreate>;
  
  
  export enum DepartmentGrade {
      S = "S",
      A = "A",
      B = "B",
  }
  
  export interface DepartmentGradeRatio {
      id: number;
      department_grade: DepartmentGrade;
      s_ratio: number;
      a_ratio: number;
      b_ratio: number;
  }
  
  export interface DepartmentGradeRatioCreate {
      department_grade: DepartmentGrade;
      s_ratio: number;
      a_ratio: number;
      b_ratio: number;
  }
  
  export type DepartmentGradeRatioUpdate = Partial<DepartmentGradeRatioCreate>;
  
  
  export enum UserRole {
      EMPLOYEE = "employee",
      TEAM_LEAD = "team_lead",
      DEPT_HEAD = "dept_head",
      ADMIN = "admin",
  }
  
  export enum EvaluationItem {
      PEER_REVIEW = "peer_review",
      PM_REVIEW = "pm_review",
      QUALITATIVE_REVIEW = "qualitative_review",
  }
  
  export interface EvaluationWeight {
      id: number;
      user_role: UserRole;
      item: EvaluationItem;
      weight: number;
  }
  
  export interface EvaluationWeightCreate {
      user_role: UserRole;
      item: EvaluationItem;
      weight: number;
  }
  
  export type EvaluationWeightUpdate = Partial<EvaluationWeightCreate>;

export interface Evaluation {
    id: number;
    name: string;
}

export interface MyEvaluationResult {
    evaluation_period: string;
    grade: string | null;
    pm_scores: {
        project_name: string;
        pm_name: string;
        score: number;
    }[];
}

export interface PeerEvaluation {
    id: number;
    project_id: number;
    evaluator_id: number;
    evaluatee_id: number;
    scores: number[];
    comment: string | null;
}

export interface PmEvaluation {
    id: number;
    project_id: number;
    evaluator_id: number;
    evaluatee_id: number;
    score: number;
    comment: string | null;
}

export interface QualitativeEvaluation {
    id: number;
    evaluator_id: number;
    evaluatee_id: number;
    score: number;
    comment: string | null;
}

export interface Evaluation {
    id: number;
    name: string;
}

export interface MyEvaluationResult {
    evaluation_period: string;
    grade: string | null;
    pm_scores: {
        project_name: string;
        pm_name: string;
        score: number;
    }[];
}

export interface PeerEvaluation {
    id: number;
    project_id: number;
    evaluator_id: number;
    evaluatee_id: number;
    scores: number[];
    comment: string | null;
}

export interface PmEvaluation {
    id: number;
    project_id: number;
    evaluator_id: number;
    evaluatee_id: number;
    score: number;
    comment: string | null;
}

export interface QualitativeEvaluation {
    id: number;
    evaluator_id: number;
    evaluatee_id: number;
    score: number;
    comment: string | null;
}

export interface PeerEvaluationCreate {
    project_id: number;
    evaluatee_id: number;
    scores: number[];
    comment?: string | null;
}
export interface PeerEvaluationUpdate extends PeerEvaluationCreate {}

export interface PmEvaluationCreate {
    project_id: number;
    evaluatee_id: number;
    score: number;
    comment?: string | null;
}
export interface PmEvaluationUpdate extends PmEvaluationCreate {}

export interface QualitativeEvaluationUpdate {
    evaluatee_id: number;
    score: number;
    comment?: string | null;
}

export interface FinalEvaluation {
  evaluatee_id: number;
  evaluation_period: string;
  peer_score: number;
  pm_score: number;
  qualitative_score: number;
  final_score: number;
  grade: string | null;
  id: number;
}

export type FinalEvaluationHistory = FinalEvaluation;

export interface PeerFeedback {
    id: number;
    evaluator_id: number;
    evaluatee_id: number;
    project_id: number;
    evaluation_period: string;
    score: number;
    feedback: string;
}

export interface ManagerEvaluationView {
    final_evaluation: FinalEvaluation | null;
    peer_feedback: PeerFeedback[];
}

export interface GradeAdjustment {
  user_id: number;
  adjusted_grade: string;
}



export interface EvaluationPeriod {

  id: number;

  name: string;

  start_date: string;

  end_date: string;

}



export interface EvaluatedUser {

  user_id: number;

  full_name: string;

  title: string;

  organization_name: string;

}



export interface DetailedProjectEvaluation {

  project_id: number;

  project_name: string;

  participation_weight: number;

  peer_evaluation_score: number | null;

  pm_evaluation_score: number | null;

  peer_feedback: string[];

}



export interface DetailedQualitativeEvaluation {

  score: number | null;

  comment: string | null;

}



export interface DetailedEvaluationResult {

  status: 'COMPLETED' | 'IN_PROGRESS';

  user_info: {

    user_id: number;

    full_name: string;

    title: string;

    organization_name: string;

  };

  final_evaluation: {

    grade: string | null;

    final_score: number | null;

  } | null;

  project_evaluations: DetailedProjectEvaluation[];

  qualitative_evaluation: DetailedQualitativeEvaluation | null;

}





// For submitting evaluations (POST)
export interface PeerEvaluationSubmit {
    evaluations: {
        project_id: number;
        evaluatee_id: number;
        scores: number[];
        comment?: string;
    }[];
}

export interface PmEvaluationSubmit {
    evaluations: {
        project_id: number;
        evaluatee_id: number;
        score: number;
        comment?: string;
    }[];
}

export interface QualitativeEvaluationCreate {
    evaluations: {
        evaluatee_id: number;
        score: number;
    }[];
}

// For fetching evaluation data (GET)
interface PeerToEvaluate {
    evaluatee_id: number;
    evaluatee_name: string;
    scores: (number | null)[];
    comment: string | null;
}



export interface PeerEvaluationData {

    project_id: number;

    project_name: string;

    status: 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED';

    peers_to_evaluate: PeerToEvaluate[];

}







// Assuming PM evaluation data has a similar structure



export interface PmEvaluationData {



    project_id: number;



    project_name: string;



    status: 'NOT_STARTED' | 'IN_PROGRESS' | 'COMPLETED';



    members_to_evaluate: MemberToEvaluate[];



}







interface MemberToEvaluate {



    evaluatee_id: number;



    evaluatee_name: string;



    score: number | null;



    comment: string | null;



}







export interface MyEvaluationTask {



    project_id: number;



    project_name: string;



    user_role_in_project: 'PM' | 'MEMBER';



}




